### PR TITLE
NEXUS-17451 - add optional limit request parameter to REST APIs

### PR DESCRIPTION
**Pull request by ruckc**
for assets, components, and search

### CODE REVIEW

**Comment by peterlynch** [comment]
One of the requirements of the issue is that a maximum limit be enforced.

This PR appears to allow changing the limit to whatever is the whim of the caller. For that reason I don't think we accept this PR before such a max limit is enforced.

Unfortunately I don't think we have identified a max limit we are comfortable with.

**Comment by ruckc** [comment]
I imagined the max limit should be a configuration setting, though adding a configuration setting seems to be non-trivial as it would most likely need to be set in the UI somewhere, though the UI doesn't have global configuration settings anywhere for something like this.

**Comment by the-player777** [comment]
Any update on this PR? We also have a lot of components and this pagination force us to to hundreds of calls to retrieve them

**Comment by icecuber** [comment]
> Any update on this PR? We also have a lot of components and this pagination force us to to hundreds of calls to retrieve them

I have to deal with quite big repository, i'm going to call **/v1/components** endpoint almost **5 THOUSAND** times 
as low as 100 items per call would be a great improvement!

**Comment by FrancisLevesque** [comment]
@peterlynch @ruckc We're sitting in the same boat. We have several scripts that regularly use the Nexus API to do various tasks and this change would dramatically cut down their running time and impact on the network.

**Comment by axi92** [comment]
We have the same problem, I have to run about 1800 requests to get all our assets from one repo.
Only the assets to get takes about 30min.

**Comment by qasimmohammed** [comment]
This would be a great addition, is there any update to this?

**Comment by joedragons** [comment]
If you haven't, suggest putting weight (comments, votes) on https://issues.sonatype.org/browse/NEXUS-17451 itself.  I don't know how regularly (if at all) leadership looks at these PRs except when weighing interest.

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/rest/internal/resources/AssetsResource.java [hunk 1]
```diff
@@ -98,13 +99,14 @@ public AssetsResource(final BrowseService browseService,
 
   @GET
   public Page<AssetXO> getAssets(@QueryParam("continuationToken") final String continuationToken,
-      @QueryParam("repository") final String repositoryId)
+      @QueryParam("repository") final String repositoryId,
+      @QueryParam("limit") @DefaultValue("10") final int limit)
```
Comment (BobbyJohansen) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/rest/internal/resources/AssetsResource.java:103]:
```suggestion
      @QueryParam("limit") @DefaultValue("10") @Max(500) final int limit)
```

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/rest/internal/resources/ComponentsResource.java [hunk 2]
```diff
@@ -130,14 +131,15 @@ public ComponentsResource(final RepositoryManagerRESTAdapter repositoryManagerRE
   @Override
   @GET
   public Page<ComponentXO> getComponents(@QueryParam("continuationToken") final String continuationToken,
-                                         @QueryParam("repository") final String repositoryId)
+                                         @QueryParam("repository") final String repositoryId,
+                                         @QueryParam("limit") @DefaultValue("10") final int limit)
```
Comment (BobbyJohansen) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/rest/internal/resources/ComponentsResource.java:135]:
```suggestion
                                         @QueryParam("limit") @DefaultValue("10") @Max(500) final int limit)
```

**Comment by RuanXinyu** [comment]
这是来自QQ邮箱的假期自动回复邮件。您好，我最近正在休假中，无法亲自回复您的邮件。我将在假期结束后，尽快给您回复。

**Comment by Za7BD** [comment]
Что-то мне подсказывает, что не дождемся) Придется костылить.

**Comment by jwaffe75** [comment]
Hello, something like this functionality would be a big help for us too.

**Comment by RuanXinyu** [comment]
这是来自QQ邮箱的假期自动回复邮件。您好，我最近正在休假中，无法亲自回复您的邮件。我将在假期结束后，尽快给您回复。

**Comment by zhl-tcsec** [comment]
In favor of

**Comment by rageofgods** [comment]
great addition. any updates?

**Comment by hybby** [comment]
Agreed that this would be an excellent addition to the API. A hard-coded 10 items per response is very low...

**Comment by RuanXinyu** [comment]
这是来自QQ邮箱的假期自动回复邮件。您好，我最近正在休假中，无法亲自回复您的邮件。我将在假期结束后，尽快给您回复。

**Comment by mhashim6** [comment]
@BobbyJohansen Any news on this one?

**Comment by baryluk** [comment]
I would also like to have this. The default value is extremely low, causing a really slow listing. We have 20k assets. I would like to use something like page size of 200, instead of 10.

**Comment by RuanXinyu** [comment]
这是来自QQ邮箱的假期自动回复邮件。您好，我最近正在休假中，无法亲自回复您的邮件。我将在假期结束后，尽快给您回复。

**Comment by tuxillo** [comment]
What happened?

**Comment by DimisNevedimis** [comment]
> What happened?

Looks like @ruckc just tired to wait almost 3 years) for Sonatype to approve and merge his PR.

Unfortunately, this is a huge problem for many of us, who has a pretty big Nexus instance.
In our case, we have almost 170k of packages that we need to index periodically. Somebody writes about 30 mins or even 1 hour of taking components and assets. Share your pain guys, in our case it takes days :(

**Comment by ruckc** [comment]
I got tired of the sheer number of security updates coming out of dependabot.  I have also given up on Sonatype making any kind of user improvements.  My company was evening a paying customer for a while and they couldn't prioritize anything even with support.

Wish everyone luck that needs this.  You might get it by 2030.

**Comment by tuxillo** [comment]
> > What happened?
> 
> Looks like @ruckc just tired to wait almost 3 years for Sonatype to approve and merge his PR.
> 
> Unfortunately, this is a huge problem for many of us, who has a pretty big Nexus instance. In our case, we have almost 170k of packages that we need to index periodically. Somebody writes about 30 mins or even 1 hour of taking components and assets. Share your pain guys, in our case it takes days. :(

What kind of repo is it? For RPM/DEB/Nuget we ended up dealing with the indexes ourselves instead of using the API ....

**Comment by tuxillo** [comment]
> I got tired of the sheer number of security updates coming out of dependabot. I have also given up on Sonatype making any kind of user improvements. My company was evening a paying customer for a while and they couldn't prioritize anything even with support.
> 
> Wish everyone luck that needs this. You might get it by 2030.

Yeah, they don't seem very responsive at all. Too bad.

**Comment by DimisNevedimis** [comment]
> > > What happened?
> > 
> > 
> > Looks like @ruckc just tired to wait almost 3 years for Sonatype to approve and merge his PR.
> > Unfortunately, this is a huge problem for many of us, who has a pretty big Nexus instance. In our case, we have almost 170k of packages that we need to index periodically. Somebody writes about 30 mins or even 1 hour of taking components and assets. Share your pain guys, in our case it takes days. :(
> 
> What kind of repo is it? For RPM/DEB/Nuget we ended up dealing with the indexes ourselves instead of using the API ....

NuGet. The reason – too slow to get data from Nexus with API, and it's much faster if you have some outer index to work with and update it periodically.

**Comment by tuxillo** [comment]
> 
> NuGet. The reason – too slow to get data from Nexus with API, and it's much faster if you have some outer index to work with and update it periodically.

Have you tried checking against the v3 indexes or if you are using v2, against the OData API (yes deprecated in nuget.org but still) ? That should be fast.

**Comment by DimisNevedimis** [comment]
> > NuGet. The reason – too slow to get data from Nexus with API, and it's much faster if you have some outer index to work with and update it periodically.
> 
> Have you tried checking against the v3 indexes or if you are using v2, against the OData API (yes deprecated in nuget.org but still) ? That should be fast.

Unfortunately, Nexus does not realize a catalogue API endpoint to list all packages. As far as I know, only nuget gallery on nuget.org have such thing.

**Pull request closed without merging**